ciao-controller/cnci.go

```diff
@@ -0,0 +1,459 @@
+func (c *CNCIManager) WaitForActive(subnet int) error {
+	c.cnciLock.Lock()
+
+	cnci, ok := c.subnets[subnet]
+	if ok {
+		// subnet already exists
+		c.cnciLock.Unlock()
+
+		// block until subnet is active
+		return c.waitForActive(subnet)
+	}
+
+	cnci = &CNCI{
+		ctrl:    c.ctrl,
+		eventCh: make(chan event),
+		subnet:  subnet,
+	}
+
+	c.subnets[subnet] = cnci
+
+	subnetBytes := make([]byte, 2)
+	binary.BigEndian.PutUint16(subnetBytes, uint16(subnet))
+	ip := net.IPv4(172, subnetBytes[0], subnetBytes[1], 0)
+	ipNet := net.IPNet{
+		IP:   ip,
+		Mask: net.IPv4Mask(255, 255, 255, 0),
+	}
+	subnetStr := ipNet.String()
+
+	// send a launch command
+	instance, err := c.launch(subnetStr)
+	if err != nil {
+		c.cnciLock.Unlock()
+		return err
+	}
+
+	glog.V(2).Infof("AddSubnet CNCI instance is %s", instance.ID)
+
+	cnci.instance = instance
+	cnci.subnet = subnet
+
+	c.cncis[instance.ID] = cnci
+
+	c.cnciLock.Unlock()
+
+	// we release the lock before waiting because
+	// we need to be able to read the event channel.
+	return cnci.waitForEventTimeout(added, cnciEventTimeout)
+}
+
+// RemoveSubnet is called when a subnet no longer is needed.
+// a cnci can be stopped.
+func (c *CNCIManager) RemoveSubnet(subnet int) error {
+	c.cnciLock.Lock()
+
+	cnci, ok := c.subnets[subnet]
+	if !ok {
+		// there is no cnci for this subnet
+		c.cnciLock.Unlock()
+		return errors.New("Subnet doesn't exist")
+	}
+
+	delete(c.subnets, subnet)
+
+	err := cnci.stop()
+	if err != nil {
+		c.cnciLock.Unlock()
+		return err
+	}
+
+	c.cnciLock.Unlock()
+
+	err = cnci.waitForEventTimeout(removed, cnciEventTimeout)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ConcentratorInstanceRemoved will move the CNCI back to the initial state
+// and send an event through the event channel.
+func (c *CNCIManager) ConcentratorInstanceRemoved(id string) error {
+	c.cnciLock.Lock()
+	defer c.cnciLock.Unlock()
+
+	cnci, ok := c.cncis[id]
+	if !ok {
+		return errors.New("No CNCI found")
+	}
+
+	cnci.transitionState(exited)
+
+	delete(c.cncis, cnci.instance.ID)
+
+	return nil
+}
+
+// ConcentratorInstanceAdded will move the CNCI into the active state
+// and send an event through the event channel.
+func (c *CNCIManager) ConcentratorInstanceAdded(id string) error {
+	c.cnciLock.Lock()
+	defer c.cnciLock.Unlock()
+
+	cnci, ok := c.cncis[id]
+	if !ok {
+		return errors.New("No CNCI found")
+	}
+
+	cnci.transitionState(active)
+
+	return nil
+}
+
+// StartFailure will move the CNCI to the error state and
+// send an event through the event channel.
+func (c *CNCIManager) StartFailure(id string) error {
+	c.cnciLock.Lock()
+	defer c.cnciLock.Unlock()
+
+	cnci, ok := c.cncis[id]
+	if !ok {
+		return errors.New("No CNCI found")
+	}
+
+	cnci.transitionState(failed)
+
+	// we should probably not do this, and instead we should
+	// delete from the map?
+	cnci.instance = nil
+
+	return nil
+}
+
+func (c *CNCIManager) waitForActive(subnet int) error {
```

Comment:
Please rethink your use of channels. To do the multiple reader on the same channel you have two options:

a.) Close the channel. All readers will unblock and move forward. Removing the need for all the polling stuff below. We have code just like this in Ciao already at https://github.com/01org/ciao/blob/master/ciao-controller/command.go#L140

Obviously by closing a channel you can only convey completion.

b.) If you want to send the information down the channel then store a slice of channels in the struct (protected by a lock) and have everyone who wants to listen put their channel in the struct and then they can just wait for the answer on their own channel.

c.) Or using a condition variable: https://golang.org/pkg/sync/#Cond
No need for polling in either case.